ESQL: Fix error on sorting unsortable geo_point and cartesian_point

craigtaverner opened the pull request:
Without a LIMIT the correct error worked, but with LIMIT it did not. This fix mimics the same error with LIMIT and adds tests for all three scenarios:

* Without limit
* With Limit
* From row with limit

Fixes #106007

craigtaverner committed:
Fix error on sorting unsortable geo_point and cartesian_point

Without a LIMIT the correct error worked, but with LIMIT it did not. This fix mimics the same error with LIMIT and adds tests for all three scenarios:
* Without limit
* With Limit
* From row with limit

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Update docs/changelog/106351.yaml

craigtaverner committed:
Add tests for geo_shape and cartesian_shape also

craigtaverner committed:
Merge branch 'fix106007' of github.com:craigtaverner/elasticsearch into fix106007

craigtaverner committed:
Updated changelog

craigtaverner committed:
Separate point and shape error messages

astefan reviewed (changes requested):
I don't believe this is the best place for such a check. Have a look at the `Verifier` for this instead.

Also, it's not clear to me (due to lack of geo knowledge), `geo_distance` is an ES specific query. Having this appear as an ESQL error message is not right. When I created the issue I wrongly assumed the error `can't sort on geo_point or point field without using specific sorting feature, like geo_distance` comes from ESQL, but it's ES itself.

craigtaverner committed:
Move error to later so we get it only if geo field is actually used in sort.

craigtaverner committed:
Implemented planner check in Verifier instead

This is a much better solution.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -381,6 +384,19 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
 
+    /**
+     * Makes sure that spatial types do not appear in sorting contexts.
+     * This check is a simplification of a similar check in the SQL verifier.
```
This statement is not needed imo.

craigtaverner replied:
Removed

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -381,6 +384,19 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
+        p.forEachDown(OrderBy.class, o -> o.order().forEach(agg -> agg.forEachUp(FieldAttribute.class, fa -> {
+            DataType dataType = fa.field().getDataType();
+            if (EsqlDataTypes.isSpatial(dataType)) {
+                localFailures.add(fail(fa, "can't sort on " + dataType.typeName()));
```
Use the longer version, please. `cannot`

astefan reviewed (commented):
Please, add unit tests for a wider set of queries (sort, sort + limit, any other place where this is not supported) in `VerifierTests`. Also, use `row` commands in your unit tests.

astefan commented:
And we need a statement in Documentation -> Limitations pages, as well.

craigtaverner committed:
Revert previous solution

craigtaverner committed:
Also check non-field attributes so the same error is provided for ROW

craigtaverner committed:
Changed "can't" to "cannot"

craigtaverner committed:
Add unit tests for verifier error

craigtaverner committed:
Added sort limitations to documentation

astefan reviewed (commented):
Looks good, but add some tests with the `from` command, as well, to VerifierTests. Sorry, I should have been more explicit last time I mentioned VerifierTests.

craigtaverner committed:
Added unit tests for spatial fields in VerifierTests

craigtaverner commented:
> Looks good, but add some tests with the `from` command, as well, to VerifierTests. Sorry, I should have been more explicit last time I mentioned VerifierTests.

OK. Added those too.

astefan reviewed (approved):
LGTM. Thanks

craigtaverner committed:
Don't run the new yaml tests on older versions

These tests mostly test the validation errors which were changed only in 8.14.0, so should not be tested in earlier versions.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -381,6 +384,18 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
+     * Makes sure that spatial types do not appear in sorting contexts.
+     */
+    private static void checkForSortOnSpatialTypes(LogicalPlan p, Set<Failure> localFailures) {
+        p.forEachDown(OrderBy.class, o -> o.order().forEach(agg -> agg.forEachUp(Attribute.class, fa -> {
```
The outer loop already does forDown, so no need for you to do it. Also Order is an expression so you don't have to unroll the list; instead do:
```
if (p instanceof OrderBy ob) {
    ob.forEachExpressionOnly(Attribute, fa -> ...)
}
```

Similar to what the other methods are doing.

craigtaverner replied:
I had copied the similar code from the SQL module, so did not think about the duplicate forEachDown, but I've removed it now. The code has more lines, but is more readable and no longer has the duplicated forEachDown.

costin reviewed (approved):
Left some comments otherwise LGTM

craigtaverner committed:
Simplify check based on code review, skip duplicate forEachDown

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.13](https://github.com/elastic/elasticsearch/pull/106379)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/106379">](https://github.com/elastic/elasticsearch/pull/106379) |
